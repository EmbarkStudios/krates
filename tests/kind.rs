mod util;

use util::build;

#[test]
fn all_the_things() {
    let grafs = build("all-features.json", krates::Builder::new()).unwrap();

    assert!(grafs.filtered.is_empty());
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        // This shouldn't actually affect anything, as dev dependencies
        // for non-workspace crates are already not resolved
        assert!(grafs.filtered.is_empty());
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        // This will be equivalent to to filtering workspace dev crates
        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_normal() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build_and_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn only_b() {
    let mut kb = krates::Builder::new();
    kb.include_workspace_crates(["/home/jake/code/krates/tests/ws/b/Cargo.toml"]);

    let grafs = build("all-features.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_after_build() {
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Dev);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }

    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Build);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }
}
