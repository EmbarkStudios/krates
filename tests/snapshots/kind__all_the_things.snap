---
source: tests/kind.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 path+file:///krates/tests/ws/a" ]
    1 [ label = "crate aho-corasick 0.7.6" ]
    2 [ label = "crate anyhow 1.0.26" ]
    3 [ label = "crate b 0.1.0 path+file:///krates/tests/ws/b" ]
    4 [ label = "crate bindgen 0.51.1" ]
    5 [ label = "crate bitflags 1.2.1" ]
    6 [ label = "crate bumpalo 3.1.2" ]
    7 [ label = "crate byteorder 1.3.2" ]
    8 [ label = "crate c 0.1.0 path+file:///krates/tests/ws/c" ]
    9 [ label = "crate cc 1.0.50" ]
    10 [ label = "crate cc 1.0.84 git+https://github.com/alexcrichton/cc-rs?branch=main" ]
    11 [ label = "crate cc 1.0.84 git+https://github.com/alexcrichton/cc-rs?rev=34d4ce4" ]
    12 [ label = "crate cexpr 0.3.6" ]
    13 [ label = "crate cfg-if 0.1.10" ]
    14 [ label = "crate clang-sys 0.28.1" ]
    15 [ label = "crate coreaudio-rs 0.9.1" ]
    16 [ label = "crate coreaudio-sys 0.2.3" ]
    17 [ label = "crate difference 2.0.0" ]
    18 [ label = "crate glob 0.3.0" ]
    19 [ label = "crate heck 0.3.1" ]
    20 [ label = "crate js-sys 0.3.35" ]
    21 [ label = "crate lazy_static 1.4.0" ]
    22 [ label = "crate leftpad 0.2.0" ]
    23 [ label = "crate libc 0.2.66" ]
    24 [ label = "crate libloading 0.5.2" ]
    25 [ label = "crate log 0.4.8" ]
    26 [ label = "crate memchr 2.2.1" ]
    27 [ label = "crate nix 0.16.1" ]
    28 [ label = "crate nom 4.2.3" ]
    29 [ label = "crate peeking_take_while 0.1.2" ]
    30 [ label = "crate proc-macro2 1.0.7" ]
    31 [ label = "crate quote 1.0.2" ]
    32 [ label = "crate regex 1.3.3" ]
    33 [ label = "crate regex-syntax 0.6.13" ]
    34 [ label = "crate ring 0.16.9" ]
    35 [ label = "crate rustc-hash 1.0.1" ]
    36 [ label = "crate shlex 0.1.1" ]
    37 [ label = "crate sourcefile 0.1.4" ]
    38 [ label = "crate spin 0.5.2" ]
    39 [ label = "crate syn 1.0.13" ]
    40 [ label = "crate thread_local 1.0.0" ]
    41 [ label = "crate unicode-segmentation 1.6.0" ]
    42 [ label = "crate unicode-xid 0.2.0" ]
    43 [ label = "crate untrusted 0.7.0" ]
    44 [ label = "crate version_check 0.1.5" ]
    45 [ label = "crate void 1.0.2" ]
    46 [ label = "crate wasm-bindgen 0.2.58" ]
    47 [ label = "crate wasm-bindgen-backend 0.2.58" ]
    48 [ label = "crate wasm-bindgen-futures 0.4.8" ]
    49 [ label = "crate wasm-bindgen-macro 0.2.58" ]
    50 [ label = "crate wasm-bindgen-macro-support 0.2.58" ]
    51 [ label = "crate wasm-bindgen-shared 0.2.58" ]
    52 [ label = "crate wasm-bindgen-webidl 0.2.58" ]
    53 [ label = "crate web-sys 0.3.35" ]
    54 [ label = "crate weedle 0.10.0" ]
    55 [ label = "crate winapi 0.2.8" ]
    56 [ label = "crate winapi 0.3.8" ]
    57 [ label = "crate winapi-i686-pc-windows-gnu 0.4.0" ]
    58 [ label = "crate winapi-x86_64-pc-windows-gnu 0.4.0" ]
    59 [ label = "feature default" ]
    60 [ label = "feature use_std" ]
    61 [ label = "feature default" ]
    62 [ label = "feature default" ]
    63 [ label = "feature runtime" ]
    64 [ label = "feature clang_6_0" ]
    65 [ label = "feature default" ]
    66 [ label = "feature default" ]
    67 [ label = "feature default" ]
    68 [ label = "feature Crypto" ]
    69 [ label = "feature Window" ]
    70 [ label = "feature verbose-errors" ]
    71 [ label = "feature default" ]
    72 [ label = "feature audio_toolbox" ]
    73 [ label = "feature audio_unit" ]
    74 [ label = "feature core_audio" ]
    75 [ label = "feature core_midi" ]
    76 [ label = "feature open_al" ]
    77 [ label = "feature default" ]
    78 [ label = "feature winerror" ]
    79 [ label = "feature errhandlingapi" ]
    80 [ label = "feature libloaderapi" ]
    81 [ label = "feature extra_traits" ]
    82 [ label = "feature default" ]
    83 [ label = "feature default" ]
    84 [ label = "feature default" ]
    85 [ label = "feature proc-macro" ]
    86 [ label = "feature default" ]
    87 [ label = "feature default" ]
    88 [ label = "feature unicode-age" ]
    89 [ label = "feature unicode-bool" ]
    90 [ label = "feature unicode-case" ]
    91 [ label = "feature unicode-gencat" ]
    92 [ label = "feature unicode-perl" ]
    93 [ label = "feature unicode-script" ]
    94 [ label = "feature unicode-segment" ]
    95 [ label = "feature ntsecapi" ]
    96 [ label = "feature wtypesbase" ]
    97 [ label = "feature default" ]
    98 [ label = "feature proc-macro" ]
    99 [ label = "feature spans" ]
    100 [ label = "feature default" ]
    101 [ label = "feature default" ]
    102 [ label = "feature default" ]
    103 [ label = "feature full" ]
    104 [ label = "feature default" ]
    105 [ label = "feature MessageEvent" ]
    106 [ label = "feature Worker" ]
    107 [ label = "feature spans" ]
    108 [ label = "feature visit" ]
    109 [ label = "feature spans" ]
    110 [ label = "feature default" ]
    111 [ label = "feature std" ]
    112 [ label = "feature std" ]
    113 [ label = "feature std" ]
    114 [ label = "feature leftpad" ]
    115 [ label = "feature leftier-strings" ]
    116 [ label = "feature lazy_static" ]
    117 [ label = "feature libloading" ]
    118 [ label = "feature gte_clang_6_0" ]
    119 [ label = "feature gte_clang_5_0" ]
    120 [ label = "feature gte_clang_4_0" ]
    121 [ label = "feature gte_clang_3_9" ]
    122 [ label = "feature gte_clang_3_8" ]
    123 [ label = "feature gte_clang_3_7" ]
    124 [ label = "feature gte_clang_3_6" ]
    125 [ label = "feature open_al" ]
    126 [ label = "feature audio_toolbox" ]
    127 [ label = "feature audio_unit" ]
    128 [ label = "feature core_audio" ]
    129 [ label = "feature core_midi" ]
    130 [ label = "feature std" ]
    131 [ label = "feature alloc" ]
    132 [ label = "feature std" ]
    133 [ label = "feature unicode-segment" ]
    134 [ label = "feature unicode-script" ]
    135 [ label = "feature unicode-perl" ]
    136 [ label = "feature unicode-gencat" ]
    137 [ label = "feature unicode-case" ]
    138 [ label = "feature unicode-bool" ]
    139 [ label = "feature unicode-age" ]
    140 [ label = "feature unicode" ]
    141 [ label = "feature thread_local" ]
    142 [ label = "feature std" ]
    143 [ label = "feature perf-literal" ]
    144 [ label = "feature aho-corasick" ]
    145 [ label = "feature memchr" ]
    146 [ label = "feature perf-inline" ]
    147 [ label = "feature perf-dfa" ]
    148 [ label = "feature perf-cache" ]
    149 [ label = "feature perf" ]
    150 [ label = "feature lazy_static" ]
    151 [ label = "feature dev_urandom_fallback" ]
    152 [ label = "feature alloc" ]
    153 [ label = "feature quote" ]
    154 [ label = "feature proc-macro" ]
    155 [ label = "feature printing" ]
    156 [ label = "feature parsing" ]
    157 [ label = "feature derive" ]
    158 [ label = "feature clone-impls" ]
    159 [ label = "feature std" ]
    160 [ label = "feature std" ]
    161 [ label = "feature spans" ]
    0 -> 3 [ label = "" ]
    0 -> 59 [ label = "(dev)" ]
    0 -> 59 [ label = "(build) 'cfg(target_os = \"linux\")'" ]
    1 -> 26 [ label = "" ]
    1 -> 60 [ label = "" ]
    3 -> 59 [ label = "" ]
    3 -> 11 [ label = "(build)" ]
    3 -> 61 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    3 -> 48 [ label = " 'cfg(all(target_arch = \"wasm32\", target_feature = \"atomics\"))'" ]
    3 -> 48 [ label = " 'cfg(all(target_vendor = \"xboxone\"))'" ]
    4 -> 62 [ label = "" ]
    4 -> 12 [ label = "" ]
    4 -> 13 [ label = "" ]
    4 -> 63 [ label = "" ]
    4 -> 64 [ label = "" ]
    4 -> 21 [ label = "" ]
    4 -> 29 [ label = "" ]
    4 -> 30 [ label = "" ]
    4 -> 31 [ label = "" ]
    4 -> 65 [ label = "" ]
    4 -> 35 [ label = "" ]
    4 -> 36 [ label = "" ]
    8 -> 10 [ label = "(build)" ]
    8 -> 66 [ label = " 'x86_64-apple-darwin'" ]
    8 -> 67 [ label = "(dev)" ]
    8 -> 21 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    8 -> 22 [ label = "" ]
    8 -> 23 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    8 -> 27 [ label = " 'x86_64-unknown-linux-gnu'" ]
    8 -> 38 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    8 -> 68 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    8 -> 69 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    8 -> 55 [ label = " 'cfg(target_os = \"windows\")'" ]
    10 -> 23 [ label = " 'cfg(unix)'" ]
    11 -> 23 [ label = " 'cfg(unix)'" ]
    12 -> 70 [ label = "" ]
    12 -> 71 [ label = "" ]
    14 -> 18 [ label = "" ]
    14 -> 18 [ label = "(build)" ]
    14 -> 23 [ label = "" ]
    14 -> 24 [ label = "" ]
    15 -> 62 [ label = "" ]
    15 -> 16 [ label = "" ]
    15 -> 72 [ label = "" ]
    15 -> 73 [ label = "" ]
    15 -> 74 [ label = "" ]
    15 -> 75 [ label = "" ]
    15 -> 76 [ label = "" ]
    16 -> 4 [ label = "(build)" ]
    19 -> 41 [ label = "" ]
    20 -> 77 [ label = "" ]
    24 -> 9 [ label = "(build)" ]
    24 -> 78 [ label = " 'cfg(windows)'" ]
    24 -> 79 [ label = " 'cfg(windows)'" ]
    24 -> 80 [ label = " 'cfg(windows)'" ]
    25 -> 13 [ label = "" ]
    27 -> 62 [ label = "" ]
    27 -> 9 [ label = "(build) 'cfg(target_os = \"dragonfly\")'" ]
    27 -> 13 [ label = "" ]
    27 -> 81 [ label = "" ]
    27 -> 82 [ label = "" ]
    27 -> 83 [ label = "" ]
    28 -> 26 [ label = "" ]
    28 -> 60 [ label = "" ]
    28 -> 44 [ label = "(build)" ]
    30 -> 84 [ label = "" ]
    31 -> 30 [ label = "" ]
    31 -> 85 [ label = "" ]
    32 -> 86 [ label = "" ]
    32 -> 87 [ label = "" ]
    32 -> 33 [ label = "" ]
    32 -> 88 [ label = "" ]
    32 -> 89 [ label = "" ]
    32 -> 90 [ label = "" ]
    32 -> 91 [ label = "" ]
    32 -> 92 [ label = "" ]
    32 -> 93 [ label = "" ]
    32 -> 94 [ label = "" ]
    32 -> 40 [ label = "" ]
    34 -> 9 [ label = "(build)" ]
    34 -> 21 [ label = " 'cfg(any(target_os = \"android\", target_os = \"freebsd\", target_os = \"linux\", target_os = \"netbsd\", target_os = \"openbsd\", target_os = \"solaris\"))'" ]
    34 -> 23 [ label = " 'cfg(any(target_os = \"android\", target_os = \"linux\"))'" ]
    34 -> 38 [ label = " 'cfg(all(any(target_arch = \"aarch64\", target_arch = \"arm\", target_arch = \"x86\", target_arch = \"x86_64\"), not(target_os = \"ios\")))'" ]
    34 -> 43 [ label = "" ]
    34 -> 68 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    34 -> 69 [ label = " 'cfg(all(target_arch = \"wasm32\", target_vendor = \"unknown\", target_os = \"unknown\", target_env = \"\"))'" ]
    34 -> 95 [ label = " 'cfg(target_os = \"windows\")'" ]
    34 -> 96 [ label = " 'cfg(target_os = \"windows\")'" ]
    35 -> 97 [ label = "" ]
    39 -> 30 [ label = "" ]
    39 -> 85 [ label = "" ]
    39 -> 31 [ label = "" ]
    39 -> 98 [ label = "" ]
    39 -> 84 [ label = "" ]
    40 -> 21 [ label = "" ]
    46 -> 13 [ label = "" ]
    46 -> 49 [ label = "" ]
    46 -> 99 [ label = "" ]
    47 -> 100 [ label = "" ]
    47 -> 21 [ label = "" ]
    47 -> 25 [ label = "" ]
    47 -> 101 [ label = "" ]
    47 -> 102 [ label = "" ]
    47 -> 103 [ label = "" ]
    47 -> 104 [ label = "" ]
    47 -> 51 [ label = "" ]
    48 -> 13 [ label = "" ]
    48 -> 20 [ label = "" ]
    48 -> 77 [ label = "" ]
    48 -> 105 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    48 -> 106 [ label = " 'cfg(target_feature = \"atomics\")'" ]
    49 -> 102 [ label = "" ]
    49 -> 50 [ label = "" ]
    49 -> 107 [ label = "" ]
    50 -> 101 [ label = "" ]
    50 -> 102 [ label = "" ]
    50 -> 108 [ label = "" ]
    50 -> 104 [ label = "" ]
    50 -> 47 [ label = "" ]
    50 -> 109 [ label = "" ]
    50 -> 51 [ label = "" ]
    52 -> 110 [ label = "" ]
    52 -> 19 [ label = "" ]
    52 -> 25 [ label = "" ]
    52 -> 101 [ label = "" ]
    52 -> 102 [ label = "" ]
    52 -> 103 [ label = "" ]
    52 -> 104 [ label = "" ]
    52 -> 47 [ label = "" ]
    52 -> 54 [ label = "" ]
    53 -> 110 [ label = "(build)" ]
    53 -> 20 [ label = "" ]
    53 -> 37 [ label = "(build)" ]
    53 -> 77 [ label = "" ]
    53 -> 52 [ label = "(build)" ]
    54 -> 71 [ label = "" ]
    56 -> 57 [ label = " 'i686-pc-windows-gnu'" ]
    56 -> 58 [ label = " 'x86_64-pc-windows-gnu'" ]
    111 -> 1 [ label = "" ]
    111 -> 60 [ label = "" ]
    86 -> 1 [ label = "" ]
    86 -> 111 [ label = "" ]
    112 -> 2 [ label = "" ]
    110 -> 2 [ label = "" ]
    110 -> 112 [ label = "" ]
    62 -> 5 [ label = "" ]
    100 -> 6 [ label = "" ]
    113 -> 7 [ label = "" ]
    97 -> 7 [ label = "" ]
    97 -> 113 [ label = "" ]
    114 -> 8 [ label = "" ]
    114 -> 22 [ label = "" ]
    115 -> 8 [ label = "" ]
    115 -> 114 [ label = "" ]
    116 -> 8 [ label = "" ]
    116 -> 21 [ label = "" ]
    59 -> 8 [ label = "" ]
    59 -> 114 [ label = "" ]
    63 -> 14 [ label = "" ]
    63 -> 117 [ label = "" ]
    117 -> 14 [ label = "" ]
    117 -> 24 [ label = "" ]
    118 -> 14 [ label = "" ]
    119 -> 14 [ label = "" ]
    120 -> 14 [ label = "" ]
    121 -> 14 [ label = "" ]
    122 -> 14 [ label = "" ]
    123 -> 14 [ label = "" ]
    124 -> 14 [ label = "" ]
    64 -> 14 [ label = "" ]
    64 -> 124 [ label = "" ]
    64 -> 123 [ label = "" ]
    64 -> 122 [ label = "" ]
    64 -> 121 [ label = "" ]
    64 -> 120 [ label = "" ]
    64 -> 119 [ label = "" ]
    64 -> 118 [ label = "" ]
    125 -> 15 [ label = "" ]
    125 -> 76 [ label = "" ]
    66 -> 15 [ label = "" ]
    66 -> 126 [ label = "" ]
    66 -> 127 [ label = "" ]
    66 -> 128 [ label = "" ]
    66 -> 125 [ label = "" ]
    66 -> 129 [ label = "" ]
    129 -> 15 [ label = "" ]
    129 -> 75 [ label = "" ]
    128 -> 15 [ label = "" ]
    128 -> 74 [ label = "" ]
    127 -> 15 [ label = "" ]
    127 -> 73 [ label = "" ]
    126 -> 15 [ label = "" ]
    126 -> 72 [ label = "" ]
    76 -> 16 [ label = "" ]
    75 -> 16 [ label = "" ]
    74 -> 16 [ label = "" ]
    73 -> 16 [ label = "" ]
    72 -> 16 [ label = "" ]
    67 -> 17 [ label = "" ]
    130 -> 23 [ label = "" ]
    81 -> 23 [ label = "" ]
    82 -> 23 [ label = "" ]
    82 -> 130 [ label = "" ]
    60 -> 26 [ label = "" ]
    87 -> 26 [ label = "" ]
    87 -> 60 [ label = "" ]
    70 -> 28 [ label = "" ]
    70 -> 131 [ label = "" ]
    132 -> 28 [ label = "" ]
    132 -> 131 [ label = "" ]
    132 -> 60 [ label = "" ]
    71 -> 28 [ label = "" ]
    71 -> 132 [ label = "" ]
    131 -> 28 [ label = "" ]
    85 -> 30 [ label = "" ]
    101 -> 30 [ label = "" ]
    101 -> 85 [ label = "" ]
    98 -> 31 [ label = "" ]
    98 -> 85 [ label = "" ]
    102 -> 31 [ label = "" ]
    102 -> 98 [ label = "" ]
    133 -> 32 [ label = "" ]
    133 -> 94 [ label = "" ]
    134 -> 32 [ label = "" ]
    134 -> 93 [ label = "" ]
    135 -> 32 [ label = "" ]
    135 -> 92 [ label = "" ]
    136 -> 32 [ label = "" ]
    136 -> 91 [ label = "" ]
    137 -> 32 [ label = "" ]
    137 -> 90 [ label = "" ]
    138 -> 32 [ label = "" ]
    138 -> 89 [ label = "" ]
    139 -> 32 [ label = "" ]
    139 -> 88 [ label = "" ]
    140 -> 32 [ label = "" ]
    140 -> 139 [ label = "" ]
    140 -> 138 [ label = "" ]
    140 -> 137 [ label = "" ]
    140 -> 136 [ label = "" ]
    140 -> 135 [ label = "" ]
    140 -> 134 [ label = "" ]
    140 -> 133 [ label = "" ]
    141 -> 32 [ label = "" ]
    141 -> 40 [ label = "" ]
    142 -> 32 [ label = "" ]
    143 -> 32 [ label = "" ]
    143 -> 144 [ label = "" ]
    143 -> 145 [ label = "" ]
    146 -> 32 [ label = "" ]
    147 -> 32 [ label = "" ]
    148 -> 32 [ label = "" ]
    148 -> 141 [ label = "" ]
    149 -> 32 [ label = "" ]
    149 -> 148 [ label = "" ]
    149 -> 147 [ label = "" ]
    149 -> 146 [ label = "" ]
    149 -> 143 [ label = "" ]
    145 -> 32 [ label = "" ]
    145 -> 26 [ label = "" ]
    65 -> 32 [ label = "" ]
    65 -> 142 [ label = "" ]
    65 -> 149 [ label = "" ]
    65 -> 140 [ label = "" ]
    144 -> 32 [ label = "" ]
    144 -> 1 [ label = "" ]
    94 -> 33 [ label = "" ]
    93 -> 33 [ label = "" ]
    92 -> 33 [ label = "" ]
    91 -> 33 [ label = "" ]
    90 -> 33 [ label = "" ]
    89 -> 33 [ label = "" ]
    88 -> 33 [ label = "" ]
    150 -> 34 [ label = "" ]
    150 -> 21 [ label = "" ]
    151 -> 34 [ label = "" ]
    151 -> 150 [ label = "" ]
    61 -> 34 [ label = "" ]
    61 -> 152 [ label = "" ]
    61 -> 151 [ label = "" ]
    152 -> 34 [ label = "" ]
    108 -> 39 [ label = "" ]
    153 -> 39 [ label = "" ]
    153 -> 31 [ label = "" ]
    154 -> 39 [ label = "" ]
    154 -> 85 [ label = "" ]
    154 -> 98 [ label = "" ]
    155 -> 39 [ label = "" ]
    155 -> 153 [ label = "" ]
    156 -> 39 [ label = "" ]
    103 -> 39 [ label = "" ]
    157 -> 39 [ label = "" ]
    104 -> 39 [ label = "" ]
    104 -> 157 [ label = "" ]
    104 -> 156 [ label = "" ]
    104 -> 155 [ label = "" ]
    104 -> 158 [ label = "" ]
    104 -> 154 [ label = "" ]
    158 -> 39 [ label = "" ]
    84 -> 42 [ label = "" ]
    159 -> 45 [ label = "" ]
    83 -> 45 [ label = "" ]
    83 -> 159 [ label = "" ]
    160 -> 46 [ label = "" ]
    161 -> 46 [ label = "" ]
    161 -> 99 [ label = "" ]
    77 -> 46 [ label = "" ]
    77 -> 161 [ label = "" ]
    77 -> 160 [ label = "" ]
    109 -> 47 [ label = "" ]
    99 -> 49 [ label = "" ]
    99 -> 107 [ label = "" ]
    107 -> 50 [ label = "" ]
    107 -> 109 [ label = "" ]
    106 -> 53 [ label = "" ]
    69 -> 53 [ label = "" ]
    105 -> 53 [ label = "" ]
    68 -> 53 [ label = "" ]
    96 -> 56 [ label = "" ]
    78 -> 56 [ label = "" ]
    95 -> 56 [ label = "" ]
    80 -> 56 [ label = "" ]
    79 -> 56 [ label = "" ]
}

